utils/das_data_sorter.py

```diff
@@ -1,76 +1,170 @@
+            if time[i] > previous_time:
+                # if true, then time at time[i] is the next second/minute
+                if previous_time != 0:
+                    # appends index array into universal array, but ignores the first iteration
+                    universal_array.append(index_array)
+
+                previous_time = ceil(time[i])
+                index_array = [] # recreates new array if previous time has changed
+            index_array.append(i) 
+
+        # pushes the final array when for loop has completed
+        universal_array.append(index_array)
+
+        return universal_array
+
+    def mean(self, data_array:list) -> int:
+        '''Finds the average of a given set of numbers. 
+
+        Disregards None and zero in calculation. Otherwise, if input is invalid, then raises error.
+        '''
+        total = 0
+        length = len(data_array)
+        for number in data_array:
+            if number == 0 or pd.isna(number):
+                # None and zeroes are ignored
+                length -= 1
+                continue
+            elif number > 0 or number < 0:
+                # 'number' value is valid
+                total += number
+            else:
+                # 'number' value is invalid
+                raise ValueError("Data point invalid and is neither zero nor None. The data point was "+str(number)+". ("+str(type(number))+")")
+        
+        try:
+            return total/length
+        except ZeroDivisionError:
+            # in the event where length = 0, due to all the elements in data_array being ignored
+            return 0 
+
+    def average_data(self, data:D) -> D:
+        '''Returns a column of new data points that has been averaged, based on specified time unit. 
+
+        Finds the average of all data points within the same time interval. Appends this new value into an array, then continues
+        to the next time interval. Once all time intervals have been accounted for, returns the set of new data points.
+        '''
+        avg_data = []
+
+        for index_array in self.indexes:
+            avg = self.mean(data[index_array])
+            avg = round(avg, ndigits=2)
+            avg_data.append(avg)
+
+        return avg_data
+
+    def location_data(self, data:D) -> D:
+        '''Handles the location data and returns the data point at the first element of each time interval. 
+
+        gps_location requires a separate function, because the values are strings, whereas the other data columns are 
+        integers/floats. Handles it so that it will always return the location data at the first element of each time interval.
+        '''
+        new_location_data = []
+        for index_array in self.indexes:
+            first_element = index_array[0] 
+            new_location = data[first_element]
+            new_location_data.append(new_location)
+        
+        return new_location_data
+    
+    def print_output_file(self) -> None:
```

Comment:
We "print" something when it appears on the console, but when we're putting stuff into a file, "write" might be more clear

Reply:
As a reader of this code, I thought this function would print the whole output file. But once I looked at the function contents, it does something else. Try describe this function in a few words about what it does. That should help you in naming this function!